### ISSUE TITLE
Add console channel

### ISSUE THREAD

**Issue by ChristopherHaws** [hypherionmc/sdlink#28]
With DiscordSRV on spigot you could assign a specific discord channel to contain console messages and you can execute commands as console from the discord channel. I would love this feature to be added to SDLink! :)

https://docs.discordsrv.com/config/#DiscordConsoleChannelId

**Comment by hypherionmc** [comment]
The console messages part has been suggested before, I am just a bit concerned about Discord ratelimits on the sending of messages. I've seen some console logs and they are scary. 

That has to either be implemented in such a way that the messages are sent in batches, OR, that it only sends new messages after the server has already started.

The commands part has also been on the table for a while. I'm looking at a way to link discord commands back to MC commands while keeping up with perms etc, but yes, it's definitely going to be added in a future release, once I'm confident with the implementation

**Comment by akaep** [comment]
Thanks for the work on this, it's really great. This is the only feature I miss from DSRV. Having the console available in its own channel on Discord that's accessible anywhere and by specific people. Super helpful for monitoring security / breakfix. Just out of curiosity, are you worried about Discord's ability to ingest the data from console? Or is there a reason the rate limit would be approached or tested by Fabric, but not Paper/Spigot? Thanks!

**Comment by hypherionmc** [comment]
My concern is because of the Forge/Fabric/Quilt abuses the console. I've had a couple of issue reports with a mod using `particle` commands to spawn particles, and sdlink would completely lock up the server because of how many times the command was executed, when discord ratelimits got triggered.

Paper/Spigot consoles are a lot cleaner and don't log as much. So it's more a case of implementing this in a way that it doesn't

A) Dump the entire log at once (which would cross the 2000 character limit)
B) Send 100's of messages in a minute, which would trigger the rate-limit

**Comment by akaep** [comment]
Oh that makes a lot of sense. Are there different designations for types of console output? Or is it all or nothing?

**Comment by hypherionmc** [comment]
So, Both Console Relaying and Discord To MC commands have been implemented. So far, both seem to be working as intended.

Here is a sample of the console channel... As I feared, it spamming the hell out of the channel, even with a delay, and the channel relay only starting after the "Server Starting..." message has fired.

```[22:30:32] [Server thread/INFO] (DedicatedServer) Starting minecraft server version 1.19.1
[22:30:32] [Server thread/INFO] (DedicatedServer) Loading properties
[22:30:32] [Server thread/INFO] (DedicatedServer) Default game type: SURVIVAL
[22:30:32] [Server thread/INFO] (MinecraftServer) Generating keypair
[22:30:32] [Server thread/INFO] (DedicatedServer) Starting Minecraft server on:25565*
[22:30:32] [Server thread/INFO] (ServerConnectionListener) Using default channel type
[22:30:32] [Server thread/INFO] (DedicatedServer) Preparing level "world"
[22:30:33] [Server thread/INFO] (MinecraftServer) Preparing start region for dimension minecraft:overworld
[22:30:38] [Worker-Main-5/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-1/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-5/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-7/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-7/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-1/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-5/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-7/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-5/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-1/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:38] [Worker-Main-7/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:39] [Worker-Main-1/INFO] (LoggerChunkProgressListener) Preparing spawn area: 0%
[22:30:39] [Worker-Main-5/INFO] (LoggerChunkProgressListener) Preparing spawn area: 13%
[22:30:40] [Worker-Main-4/INFO] (LoggerChunkProgressListener) Preparing spawn area: 52%
[22:30:40] [Worker-Main-4/INFO] (LoggerChunkProgressListener) Preparing spawn area: 83%
[22:30:40] [Server thread/INFO] (LoggerChunkProgressListener) Time elapsed: 7364 ms
[22:30:40] [Server thread/INFO] (DedicatedServer) Done (7.973s)! For help, type "help"
[22:31:07] [User Authenticator #1/INFO] (ServerLoginPacketListenerImpl) UUID of player HypherionSA is REDACTED
[22:31:08] [Server thread/INFO] (PlayerList) HypherionSA[/127.0.0.1:55113] logged in with entity id 174 at (-49.816027638023854, 64.0, -6.934446760754822)
[22:31:08] [Server thread/INFO] (MinecraftServer) HypherionSA joined the game
[22:31:15] [Server thread/INFO] (MinecraftServer) <HypherionSA> Hello World
[22:31:35] [Server thread/INFO] (ServerGamePacketListenerImpl) HypherionSA lost connection: Disconnected
[22:31:35] [Server thread/INFO] (MinecraftServer) HypherionSA left the game
```

I might change this to only start relaying messages once the server has fully started, OR, I'll leave it as is and see what the feedback is on it

**Comment by hypherionmc** [comment]
Here is the config sample for linking commands:

First one anyone can execute, the second one is locked behind a role.

```toml
#Execute Minecraft commands in Discord
[linkedCommands]
	#Should linked commands be enabled
	enabled = true
	#Commands to be linked
	[[linkedCommands.commands]]
		mcCommand = "say %args%"
		discordCommand = "saymc"
		discordRole = ""

	[[linkedCommands.commands]]
		mcCommand = "stop"
		discordCommand = "stopmc"
		discordRole = "MacOS"
```

Commands are executed in game as a "fake" player, so permission control is possible

**Comment by hypherionmc** [comment]
Alrighty. Test builds with the Console channel/commands and formatting are ready!

You can find them here: [Jenkins](https://ci.firstdarkdev.xyz/job/Simple%20Discord%20Link/job/Forge-Fabric-Quilt/)

And a rough guide/changelog [HERE](https://gist.github.com/hypherionmc/65c7e6d251ba64e0df1fc882c68fa4f8).

If all goes well, this will be pushed to release on Saturday

**Comment by hypherionmc** [comment]
Features were released today with SDLink 2.1

**Comment by akaep** [comment]
Amazing. Thanks for all the work on this!
